### PR TITLE
feat!: snowpark

### PR DESCRIPTION
**Pull request by tobymao**

### CODE REVIEW

File: sqlmesh/core/engine_adapter/snowflake.py [hunk 1]
```diff
@@ -72,28 +75,28 @@ def session(self, properties: SessionProperties) -> t.Iterator[None]:
...
         temp_table = self._get_temp_table(target_table or "pandas")
 
         def query_factory() -> Query:
-            from snowflake.connector.pandas_tools import write_pandas
-
-            if not self.table_exists(temp_table):
-                # Workaround for https://github.com/snowflakedb/snowflake-connector-python/issues/1034
-                #
-                # The above issue has already been fixed upstream, but we keep the following
-                # line anyway in order to support a wider range of Snowflake versions.
```
Comment (eakmanrq) [inline sqlmesh/core/engine_adapter/snowflake.py:92]: If you are going to drop support for older versions, we should update the min version on Snowflake then right to reflect that?

Comment (tobymao) [reply sqlmesh/core/engine_adapter/snowflake.py:92]: this bug only affects snowflake users on old versions who happen to also use pandas. i could add a min but didn't think it was that necessary

Comment (tobymao) [reply sqlmesh/core/engine_adapter/snowflake.py:92]: after reading the dbt issues related to this, i think i'll keep the workaround for now

**Review by eakmanrq — APPROVED** [review]

File: sqlmesh/core/engine_adapter/base.py [hunk 2]
```diff
@@ -146,14 +147,14 @@ def cursor(self) -> t.Any:
...
+    @property
+    def snowpark(self) -> t.Optional[SnowparkSession]:
+        return None
+
     @property
     def comments_enabled(self) -> bool:
         return self._register_comments and self.COMMENT_CREATION_TABLE.is_supported
 
-    @classmethod
-    def is_pandas_df(cls, value: t.Any) -> bool:
```
Comment (izeigerman) [inline sqlmesh/core/engine_adapter/base.py:154]: Why are we getting rid of this?

File: sqlmesh/core/engine_adapter/base.py [hunk 3]
```diff
@@ -53,6 +53,7 @@
         PySparkSession,
         Query,
         QueryOrDF,
+        SnowparkSession,
```
Comment (izeigerman) [inline sqlmesh/core/engine_adapter/base.py:56]: I don't understand how this import can possible work with Python 3.12

File: sqlmesh/core/engine_adapter/snowflake.py [hunk 4]
```diff
@@ -72,22 +75,30 @@ def session(self, properties: SessionProperties) -> t.Iterator[None]:
         yield
         self.execute(f"USE WAREHOUSE {current_warehouse_sql}")
 
+    @property
```
Comment (izeigerman) [inline sqlmesh/core/engine_adapter/snowflake.py:78]: Should we cache this?

File: sqlmesh/core/engine_adapter/base.py (hunk 2 above)
Comment (tobymao) [reply sqlmesh/core/engine_adapter/base.py:154]: because it was confusing mypy

File: sqlmesh/core/engine_adapter/base.py (hunk 3 above)
Comment (tobymao) [reply sqlmesh/core/engine_adapter/base.py:56]: because it's an if else statement in the typing

Comment (tobymao) [reply sqlmesh/core/engine_adapter/base.py:56]: ah i see, you're right, i don't know how it works

**Pull request merged**